martypy/ClientMV2.py

```diff
@@ -400,6 +401,81 @@ def preException(self, isFatal: bool) -> None:
         logger.debug(f"Pre-exception isFatal {isFatal}")
+        
+    def valid_addon(self, add_on:str) -> bool :
+        disco= {"00000087","00000088","00000089"}
+        for add_ons in self.get_add_ons_status().values():
+            if add_ons['name'] == add_on:
+                if add_ons['whoAmITypeCode'] in disco:
+                    return True
+                else:
+                    raise MartyCommandException("The add on name passed in is not a valid disco add on.")
+        raise MartyCommandException("The add on name passed in is not a valid add on.")
+
+    def valid_hex(self, hexc: str) -> bool :
+        regex = "^([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$"
+        p = re.compile(regex)
+        if not hexc:
+            return False
+        if (re.search(p,hexc)):
+            return True
+        return False
+
+    def disco_off(self, add_on: str = 'all') -> bool :
+        pattern = '01'
+        if self.valid_addon(add_on):
+            return self.ricIF.cmdRICRESTRslt(f"elem/{add_on}/json?cmd=raw&hexWr={pattern}")
+
+    def disco_pattern(self,  pattern: str, add_on: str ) -> bool :
+        if self.valid_addon(add_on):
+            return self.ricIF.cmdRICRESTRslt(f"elem/{add_on}/json?cmd=raw&hexWr={pattern}")
+
+    def disco_cmd_hex(self, hexc: str, add_on: str, region: int) -> bool:
+        if region == 'all':
+            region = '02'
+        else:
+            region = '040' + str(region)
+        if self.valid_addon(add_on):
+            return self.ricIF.cmdRICRESTRslt(f"elem/{add_on}/json?cmd=raw&hexWr={region}{hexc}")
+
+    def disco_color(self, color: Union[int,str,tuple], add_on: str, region: Union[int,str] = 'all') -> bool:#mayb switch  color and add_on
+        default_colors = {
+            'white':'FFFFFF',
+            'red':'FF0000',
+            'blue':'0000FF',
+            'yellow':'FFFF00',
+            'green':'008000',
+            'teal':'008080',
+            'pink':'800080',
+            'purple':'060014',
+            'orange':'0f0200'
+        }
+        if type(color) is str:
+            try:
+                hex_color = default_colors[color.lower()]
+                return self.disco_cmd_hex(hex_color, add_on,region)
+            except KeyError:
+                if color[0] == '#':
+                    color = color[1:]
+                if self.valid_hex(color): # add other params of hex code here
+                    return self.disco_cmd_hex(color, add_on,region)
+                else:
+                    raise MartyCommandException("Color specified is not a valid hex code or default color")
+        if type(color) is tuple:
+            if len(color)!= 3:
+                raise MartyCommandException("RGB tuple must be 3 numbers, please enter valid color.")
+            hex_color = '%02x%02x%02x' % color
+            return self.disco_cmd_hex(hex_color, add_on,region)
+
+    def disco_group(self, function: str, group: set = {"00000087","00000088","00000089"}, params: dict = {}) :
```

Comment:
Great idea to have this take one of the other functions as an argument! :100: I hadn't thought of that.

It would warrant a docstring though because I have no idea what the function does without reading it's implementation in detail. I'm also thinking if the naming could be improved to achieve the same.

For example, `function` could be `disco_function` or `operation`, `group` could be `whoami_type_codes` or something, and `params` could perhaps be `operation_kwargs` or `disco_fn_kwargs`. Also the method could be named something like `disco_group_operation` - I think there is an even better name out there but I can't think of it rn. :smile: 

On another note, `function` is not of type `str`. You can use [Callable](https://docs.python.org/3/library/typing.html#callable) - it even lets you specify arguments and return types. Also, `group` is always specified explicitly, isn't it? Maybe we can drop the default which would save a lot of space in the function signature. :smile:

Lastly, although I don't think it's a problem in this case, [it is usually a bad idea](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) to put mutable things as argument defaults (both `group` and `params` in this case).